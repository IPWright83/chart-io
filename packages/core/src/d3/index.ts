import * as d3 from "./barrel";
export { d3 };
