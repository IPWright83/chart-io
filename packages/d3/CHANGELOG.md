# @d3-chart/d3

## 0.3.0

### Minor Changes

- 494a204: Test
- 5c9ca11: testing

## 0.2.0

### Minor Changes

- feddec9: TypeScript support
