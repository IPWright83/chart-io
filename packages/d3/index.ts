export * from "d3-array";
export * from "d3-axis";
export * from "d3-brush";
export * from "d3-color";
export * from "d3-ease";
export * from "d3-format";
export * from "d3-selection";
export * from "d3-shape";
export * from "d3-time-format";
export * from "d3-time";
export * from "d3-timer";
export * from "d3-transition";
export * from "d3-scale";
