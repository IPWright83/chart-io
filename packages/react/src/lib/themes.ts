import type { IColor, ITheme } from "@chart-io/types";

const light: ITheme = {
    axis: {
        stroke: "#333333" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    droplines: {
        strokeDasharray: 4,
        strokeOpacity: 1,
        strokeWidth: 2,
    },

    markers: {
        size: 6,
        stroke: "white" as IColor,
        strokeWidth: 2,
        shadow: false,
    },

    font: {
        family: "sans-serif",
        size: 12,
    },

    background: "#ffffff" as IColor,

    crosshair: {
        stroke: "#000000" as IColor,
        strokeOpacity: 0.5,
        strokeWidth: 1,
        strokeDasharray: 0,
    },

    gridlines: {
        stroke: "#d3d3d3" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    tooltip: {
        background: "#ffffff" as IColor,
        text: "#000000" as IColor,
        opacity: 0.8,
        padding: 8,
    },

    legend: {
        background: "#ffffff" as IColor,
        text: "#000000" as IColor,
        opacity: 1,
        padding: 8,
        border: "#ccc",
        defaultMaxWidth: 200,
        defaultMaxHeight: 50,
    },

    series: {
        opacity: 0.7,
        selectedOpacity: 1,
        colors: [
            "#99c1dc",
            "#fc998e",
            "#fdc381",
            "#c2e587",
            "#fff18c",
            "#c999ca",
            "#a4dcd2",
            "#d6efd1",
            "#ffffc2",
            "#cbc8e1",
            "#fdd7ea",
            "#e1e1e1",
        ] as IColor[],
    },
};

const dark: ITheme = {
    axis: {
        stroke: "#cccccc" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    droplines: {
        strokeDasharray: 4,
        strokeOpacity: 1,
        strokeWidth: 2,
    },

    background: "#333333" as IColor,

    font: {
        family: "sans-serif",
        size: 12,
    },

    crosshair: {
        stroke: "#cccccc" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
        strokeDasharray: 0,
    },

    markers: {
        size: 6,
        stroke: "white" as IColor,
        strokeWidth: 2,
        shadow: false,
    },

    gridlines: {
        stroke: "#555555" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    tooltip: {
        background: "#333333" as IColor,
        text: "#ffffff" as IColor,
        opacity: 0.8,
        padding: 8,
    },

    legend: {
        background: "#333333" as IColor,
        text: "#ffffff" as IColor,
        opacity: 1,
        padding: 8,
        border: "#ccc",
        defaultMaxWidth: 200,
        defaultMaxHeight: 50,
    },

    series: {
        opacity: 1,
        selectedOpacity: 0.7,
        colors: [
            "#51cbdf",
            "#ee709d",
            "#f8c86c",
            "#677be7",
            "#90e5a4",
            "#91564b",
            "#ec73c1",
            "#7f7f7f",
            "#bdbf2f",
            "#00becf",
        ] as IColor[],
    },
};

const themes = {
    light,
    dark,
};

export { themes };
