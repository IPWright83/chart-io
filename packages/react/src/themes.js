const light = {
    axis: {
        stroke: "#333333",
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    droplines: {
        strokeDasharray: 4,
        strokeOpacity: 1,
        strokeWidth: 2,
    },

    background: "#ffffff",

    crosshair: {
        stroke: "#000000",
        strokeOpacity: 0.5,
        strokeWidth: 1,
        strokeDasharray: 0,
    },

    gridlines: {
        stroke: "#d3d3d3",
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    series: {
        opacity: 0.7,
        selectedOpacity: 1,
        colors: [
            "#99c1dc",
            "#fc998e",
            "#fdc381",
            "#c2e587",
            "#fff18c",
            "#c999ca",
            "#a4dcd2",
            "#d6efd1",
            "#ffffc2",
            "#cbc8e1",
            "#fdd7ea",
            "#e1e1e1",
        ],
    },
};

const dark = {
    axis: {
        stroke: "#cccccc",
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    droplines: {
        strokeDasharray: 4,
        strokeOpacity: 1,
        strokeWidth: 2,
    },

    background: "#333333",

    crosshair: {
        stroke: "#cccccc",
        strokeOpacity: 0.8,
        strokeWidth: 1,
        strokeDasharray: 0,
    },

    gridlines: {
        stroke: "#555555",
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    series: {
        opacity: 1,
        selectedOpacity: 0.7,
        colors: [
            "#51cbdf",
            "#ee709d",
            "#f8c86c",
            "#677be7",
            "#90e5a4",
            "#91564b",
            "#ec73c1",
            "#7f7f7f",
            "#bdbf2f",
            "#00becf",
        ],
    },
};

const themes = {
    light,
    dark,
};

export { themes };
