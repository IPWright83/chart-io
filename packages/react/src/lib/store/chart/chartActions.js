import { themes } from "../../themes";

/**
 * Sets the dimensions in the Redux store
 * @param  {Number} width   The width of the chart
 * @param  {Number} height  The height of the chart
 * @param  {Object} margin  The margin for the chart
 * @return {Object}         A redux action object
 */
const setDimensions = (width, height, margin) => {
    if (!margin) {
        // TODO: Margin is incorrect shape
        // Margin is not provided
        console.warn();
    }

    return {
        type: "CHART.SET_DIMENSIONS",
        payload: { width, height, margin },
    };
};

/**
 * Sets the scale for this field in the Redux store
 * @param  {Array<String>} fields   The names of the fields to set the scale for
 * @param  {Function} scale         A d3.Scale function
 * @param  {Boolean} fromAxis       Has this scale been autogenerated from an axis?
 * @return {Object}                 A redux action object
 */
const setScales = (fields, scale, fromAxis) => (dispatch) => {
    // TODO log this case?
    if (!scale) return;

    dispatch({
        type: "CHART.SET_SCALES",
        payload: { fields, scale, fromAxis },
    });
};

/**
 * Sets the theme for the chart in the Redux store
 * @param  {Object} theme   The theme object
 * @return {Object}                 A redux action object
 */
const setTheme = (theme) => (dispatch) => {
    if (theme === "light") {
        dispatch({
            type: "CHART.SET_THEME",
            payload: themes.light,
        });

        return;
    }

    if (theme === "dark") {
        dispatch({
            type: "CHART.SET_THEME",
            payload: themes.dark,
        });

        return;
    }

    dispatch({
        type: "CHART.SET_THEME",
        payload: theme,
    });
};

/**
 * Sets the duration for animations in the Redux store
 * @param  {Number} duration The duration in miliseconds
 * @return {Object}          A redux action object
 */
const setAnimationDuration = (duration) => ({
    type: "CHART.SET_ANIMATION_DURATION",
    payload: duration,
});

/**
 * Sets the data for the chart in the Redux store
 * @param  {Array}  data An array of the data for the chart
 * @return {Object}      A redux action object
 */
const setData = (data = []) => {
    // TODO: Enforce arrays
    return {
        type: "CHART.SET_DATA",
        payload: data,
    };
};

const chartActions = {
    setDimensions,
    setScales,
    setData,
    setTheme,
    setAnimationDuration,
};

export { chartActions };
