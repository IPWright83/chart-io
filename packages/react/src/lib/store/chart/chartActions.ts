import type { IData, ILegendItem, IMargin, IScale, ITheme } from "@chart-io/types";

import type {
    AddLegendItemAction,
    RemoveLegendItemAction,
    SetAnimationDurationAction,
    SetBrushRangeAction,
    SetBrushDimensionsAction,
    SetChartIDAction,
    SetDataAction,
    SetDimensionAction,
    SetScaleAction,
    SetScaleZoomAction,
    SetThemeAction,
} from "./types";

import { logWarning } from "../../utils";
import { themes } from "../../themes";

/**
 * Validates the the margin is correct, returning true if so
 * @param  value     The value for the margin
 * @param  side      The side of the margin we are checking [top, left, bottom, right]
 * @return           True if the margin is valid, false otherwise
 */
const validateMargin = (value: number, side: string): boolean => {
    if (value === null || value === undefined) {
        logWarning("W005", `The ${side} of the margin was not specified and is required`);
        return false;
    }

    return true;
};

/**
 * Sets the dimensions in the Redux store
 * @param  width       The width of the chart
 * @param  height      The height of the chart
 * @param  plotMargin  The margin for the plot
 * @return             A redux action object
 */
const setDimensions = (width: number, height: number, plotMargin: IMargin): SetDimensionAction => {
    if (!plotMargin) {
        logWarning("W004", "A margin was not provided but is required");
        return;
    }

    // prettier-ignore
    if (!validateMargin(plotMargin.left, "left") || 
        !validateMargin(plotMargin.top, "top") ||
        !validateMargin(plotMargin.bottom, "bottom") ||
        !validateMargin(plotMargin.right, "right")) 
    {
        return;
    }

    return {
        type: "CHART.SET_DIMENSIONS",
        payload: { width, height, margin: plotMargin },
    };
};

/**
 * Sets the margin that should be applied to the brush in the Redux store
 * @param  width       The width of the brush
 * @param  height      The height of the brush
 * @param  brushMargin The margin for the brush
 * @return             A redux action object
 */
const setBrushDimensions = (width: number, height: number, brushMargin: IMargin): SetBrushDimensionsAction => {
    // prettier-ignore
    if (!validateMargin(brushMargin.left, "left") || 
        !validateMargin(brushMargin.top, "top") ||
        !validateMargin(brushMargin.bottom, "bottom") ||
        !validateMargin(brushMargin.right, "right")) 
    {
        return;
    }

    return {
        type: "CHART.SET_BRUSH_DIMENSIONS",
        payload: {
            width,
            height,
            margin: brushMargin,
        },
    };
};

/**
 * Sets the scale for this field in the Redux store
 * @param  fields     The names of the fields to set the scale for
 * @param  scale      A d3.Scale function
 * @return            A redux action object
 */
const setScales =
    (fields: string[], scale: IScale) =>
    (dispatch): SetScaleAction => {
        if (!scale) return;

        dispatch({
            type: "CHART.SET_SCALES",
            payload: { fields, scale },
        });
    };

/**
 * Creates a range to use for Scales with a Brush enabled in the Redux store
 * @param  field      The names of the field to set the scale for
 * @param  range      The new range to use for this scale
 * @return            A redux action object
 */
const setBrushRange = (field: string, range: number[]): SetBrushRangeAction => ({
    type: "CHART.SET_BRUSH_RANGE",
    payload: { field, range },
});

/**
 * Reserves some space for the brush to be rendered on the chart
 * @param width     The amount of space that needs reserving in the width for the brush
 * @param height    The amount of space that needs reserving in the height for the brush
 * @return            A redux action object
 */
const setBrushReservedDimensions = (width: number, height: number): SetBrushReservedDimensionsAction => ({
    type: "CHART.SET_BRUSH_RESERVED_DIMENSIONS",
    payload: { width, height },
});

/**
 * Zooms a particular scale to the given domain
 * @param  field      The names of the field to zoom the scale for
 * @param  domain     The new domain for the scale
 * @return            A redux action object
 */
const setScaleZoom = (field: string, domain: number[] | Date[]): SetScaleZoomAction => ({
    type: "CHART.SET_SCALE_ZOOM",
    payload: { field, domain },
});

/**
 * Adds a Legend item to the chart
 * @param  item    The legend item
 * @return         A redux action object
 */
const addLegendItem = (item: ILegendItem): AddLegendItemAction => ({
    type: "CHART.ADD_LEGEND_ITEM",
    payload: item,
});

/**
 * Removes a Legend item from the chart
 * @param  item    The legend item
 * @return         A redux action object
 */
const removeLegendItem = (item: ILegendItem): RemoveLegendItemAction => ({
    type: "CHART.REMOVE_LEGEND_ITEM",
    payload: item,
});

/**
 * Sets the theme for the chart in the Redux store
 * @param  theme   The theme object
 * @return         A redux action object
 */
const setTheme =
    (theme: ITheme | "light" | "dark") =>
    (dispatch): SetThemeAction => {
        if (theme === "light") {
            dispatch({
                type: "CHART.SET_THEME",
                payload: themes.light,
            });

            return;
        }

        if (theme === "dark") {
            dispatch({
                type: "CHART.SET_THEME",
                payload: themes.dark,
            });

            return;
        }

        dispatch({
            type: "CHART.SET_THEME",
            payload: theme,
        });
    };

/**
 * Sets the duration for animations in the Redux store
 * @param  duration     The duration in miliseconds
 * @return              A redux action object
 */
const setAnimationDuration = (duration: number): SetAnimationDurationAction => ({
    type: "CHART.SET_ANIMATION_DURATION",
    payload: duration,
});

/**
 * Sets the data for the chart in the Redux store
 * @param  data         An array of the data for the chart
 * @return              A redux action object
 */
const setData = (data: IData = []): SetDataAction => {
    return {
        type: "CHART.SET_DATA",
        payload: data,
    };
};

/**
 * Sets a unique ID for the chart in the Redux store
 * @param chartID   The ID of the chart
 * @return          A redux action object
 */
const setChartID = (chartID: string): SetChartIDAction => {
    return {
        type: "CHART.SET_CHART_ID",
        payload: chartID,
    };
};

const chartActions = {
    setDimensions,
    setBrushDimensions,
    setScales,
    setData,
    setTheme,
    setAnimationDuration,
    addLegendItem,
    removeLegendItem,
    setBrushRange,
    setScaleZoom,
    setChartID,
    setBrushReservedDimensions,
};

export { chartActions };
