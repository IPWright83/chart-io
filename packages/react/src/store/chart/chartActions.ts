import { themes } from "../../themes";
import type { IMargin, ITheme, IData, IScaleFunction } from "../../types";

import type { SetDimensionAction, SetAnimationDurationAction, SetDataAction } from "./types";

/**
 * Sets the dimensions in the Redux store
 * @param  width   The width of the chart
 * @param  height  The height of the chart
 * @param  margin  The margin for the chart
 * @return         A redux action object
 */
const setDimensions = (width: number, height: number, margin: IMargin): SetDimensionAction => {
    if (!margin) {
        // TODO: Margin is incorrect shape
        // Margin is not provided
        console.warn();
    }

    return {
        type: "CHART.SET_DIMENSIONS",
        payload: { width, height, margin },
    };
};

/**
 * Sets the scale for this field in the Redux store
 * @param  fields     The names of the fields to set the scale for
 * @param  scale      A d3.Scale function
 * @param  fromAxis   Has this scale been autogenerated from an axis?
 * @return            A redux action object
 */
const setScales = (fields: string[], scale: IScaleFunction, fromAxis: boolean) => (dispatch) => {
    // TODO log this case?
    if (!scale) return;

    dispatch({
        type: "CHART.SET_SCALES",
        payload: { fields, scale, fromAxis },
    });
};

/**
 * Sets the theme for the chart in the Redux store
 * @param  theme   The theme object
 * @return         A redux action object
 */
const setTheme = (theme: ITheme | "light" | "dark") => (dispatch) => {
    if (theme === "light") {
        dispatch({
            type: "CHART.SET_THEME",
            payload: themes.light,
        });

        return;
    }

    if (theme === "dark") {
        dispatch({
            type: "CHART.SET_THEME",
            payload: themes.dark,
        });

        return;
    }

    dispatch({
        type: "CHART.SET_THEME",
        payload: theme,
    });
};

/**
 * Sets the duration for animations in the Redux store
 * @param  duration     The duration in miliseconds
 * @return              A redux action object
 */
const setAnimationDuration = (duration: number): SetAnimationDurationAction => ({
    type: "CHART.SET_ANIMATION_DURATION",
    payload: duration,
});

/**
 * Sets the data for the chart in the Redux store
 * @param  data         An array of the data for the chart
 * @return              A redux action object
 */
const setData = (data: IData = []): SetDataAction => {
    // TODO: Enforce arrays
    return {
        type: "CHART.SET_DATA",
        payload: data,
    };
};

const chartActions = {
    setDimensions,
    setScales,
    setData,
    setTheme,
    setAnimationDuration,
};

export { chartActions };
