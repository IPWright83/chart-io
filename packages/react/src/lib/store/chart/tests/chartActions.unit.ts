import * as d3 from "d3";

import { themes } from "../../../themes";
import { chartActions } from "../chartActions";

describe("chartActions", () => {
    it("setDimensions returns correct action", () => {
        const width = 1000;
        const height = 500;
        const margin = { top: 10, left: 10, right: 10, bottom: 10 };

        expect(chartActions.setDimensions(width, height, margin)).toEqual({
            type: "CHART.SET_DIMENSIONS",
            payload: { width, height, margin },
        });
    });

    it("setDimensions warns with no margin", () => {
        const width = 1000;
        const height = 500;

        const spy = jest.spyOn(console, "warn").mockImplementation();

        // @ts-expect-error: Checking runtime validation
        expect(chartActions.setDimensions(width, height)).toBeUndefined();
        expect(spy).toHaveBeenCalledWith("W004 - A margin was not provided but is required");
    });

    it("setDimensions warns with invalid margin", () => {
        const width = 1000;
        const height = 500;
        const margin = { left: 10, right: 10, bottom: 10 };

        const spy = jest.spyOn(console, "warn").mockImplementation();

        // @ts-expect-error: Checking runtime validation
        expect(chartActions.setDimensions(width, height, margin)).toBeUndefined();
        expect(spy).toHaveBeenCalledWith("W005 - The top of the margin was not specified and is required");
    });

    describe("setScales", () => {
        it("triggers dispatch correctly", () => {
            const fields = ["a", "b", "c"];
            const scale = d3.scaleLinear();
            const dispatch = jest.fn();

            chartActions.setScales(fields, scale, false)(dispatch);
            expect(dispatch).toHaveBeenCalledWith({
                type: "CHART.SET_SCALES",
                payload: { fields, scale, fromAxis: false },
            });
        });

        it("does not do anything if passed no scale", () => {
            const fields = ["a", "b", "c"];
            const dispatch = jest.fn();

            // @ts-expect-error: Checking runtime validation
            chartActions.setScales(fields, null)(dispatch);
            expect(dispatch).not.toHaveBeenCalled();
        });
    });

    describe("setData", () => {
        it("returns the correct action", () => {
            const data = [{ a: "foo" }];
            expect(chartActions.setData(data)).toEqual({
                type: "CHART.SET_DATA",
                payload: data,
            });
        });

        it("defaults to an empty dataset", () => {
            expect(chartActions.setData(undefined)).toEqual({
                type: "CHART.SET_DATA",
                payload: [],
            });
        });
    });

    it("setAnimationDuration returns the correct action", () => {
        expect(chartActions.setAnimationDuration(500)).toEqual({
            type: "CHART.SET_ANIMATION_DURATION",
            payload: 500,
        });
    });

    describe("setTheme", () => {
        it("for light theme returns correct action", () => {
            const dispatch = jest.fn();

            chartActions.setTheme("light")(dispatch);

            expect(dispatch).toHaveBeenCalledWith({
                type: "CHART.SET_THEME",
                payload: themes.light,
            });
        });

        it("for dark theme returns correct action", () => {
            const dispatch = jest.fn();

            chartActions.setTheme("dark")(dispatch);

            expect(dispatch).toHaveBeenCalledWith({
                type: "CHART.SET_THEME",
                payload: themes.dark,
            });
        });

        it("for custom theme returns correct action", () => {
            const dispatch = jest.fn();

            // @ts-expect-error: Checking runtime validation
            chartActions.setTheme({ foo: "bar" })(dispatch);

            expect(dispatch).toHaveBeenCalledWith({
                type: "CHART.SET_THEME",
                payload: { foo: "bar" },
            });
        });
    });
});
