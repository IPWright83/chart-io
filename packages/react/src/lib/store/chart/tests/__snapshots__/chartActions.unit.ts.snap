// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chartActions setBrushDimensions warns with invalid margin 1`] = `"@chart-io/react encountered an warning. W005: The top of the margin was not specified and is required. You can read more about this https://ipwright83.github.io/chart-io/?path=/docs/errors-warnings-warnings-W005."`;

exports[`chartActions setDimensions warns with invalid margin 1`] = `"@chart-io/react encountered an warning. W005: The top of the margin was not specified and is required. You can read more about this https://ipwright83.github.io/chart-io/?path=/docs/errors-warnings-warnings-W005."`;

exports[`chartActions setDimensions warns with no margin 1`] = `"@chart-io/react encountered an warning. W004: A margin was not provided but is required. You can read more about this https://ipwright83.github.io/chart-io/?path=/docs/errors-warnings-warnings-W004."`;
